Add ConverterSaildrones module for Saildrones data conversion

david11133 opened the pull request:
# PR: Add Full Support for Saildrones (Met-Ocean) in CrocoLake

## Summary

This PR **fully integrates Saildrone (Met-Ocean) data** into the CrocoLake ecosystem by introducing a dedicated converter module and CLI utility. It processes raw NetCDF files, applies CrocoLake’s standardized schema, and converts them into Parquet format—ensuring compatibility, scalability, and quality control.

---

## Key Components

### Parameters

- `params["Saildrones"]`: List of relevant raw Saildrone variables to be extracted from NetCDF files.
- `params["Saildrones2CROCOLAKE"]`: Mapping from Saildrone variable names to CrocoLake’s unified schema.

### Converter Module

- **`converterSaildrones.py`**:

  - Reads NetCDF files using `xarray`, safely manages concurrency with `Dask` and `Locks`.
  - Supports **chunked processing** of large dataframes to prevent memory overhead.
  - Handles both **PHY** and **BGC** variables with flexible mappings.
  - Performs **depth-aware alignment** of variables using a predefined `depth_map`.
  - Computes **pressure (PRES)** using the **Gibbs Seawater (GSW)** library.
  - Adds **basic QC flags** for selected variables.
  - Fills variables from multiple sensors using **backward fill** logic.
  - Skips empty/invalid rows and standardizes all outputs to CrocoLake schema.

### CLI Tool

- **`saildrones2parquet.py`**: A script to run the converter on Saildrone NetCDF data.

---

## Usage

### Convert Saildrone NetCDF to Parquet:

```bash
python saildrones2parquet.py --config
```

## Example Output (Parquet Format)


```json
{"DB_NAME":"Saildrones","PLATFORM_NUMBER":"1090.0","LATITUDE":9.3771672,"LONGITUDE":-124.946432,"JULD":1730396700000,"PRES":0.5028322339057922,"PRES_QC":null,"PRES_ERROR":null,"TEMP":28.435400009155273,"TEMP_QC":null,"TEMP_ERROR":null,"PSAL":32.72679901123047,"PSAL_QC":null,"PSAL_ERROR":null,"DOXY":200.83999633789062,"DOXY_QC":1,"DOXY_ERROR":null,"BBP700":null,"BBP700_QC":1,"BBP700_ERROR":null,"CHLA":0.05000000074505806,"CHLA_QC":1,"CHLA_ERROR":null,"CDOM":null,"CDOM_QC":1,"CDOM_ERROR":null,"ABS_SAL_COMPUTED":32.881500244140625,"CONSERVATIVE_TEMP_COMPUTED":28.537002563476562,"SIGMA1_COMPUTED":24.71753692626953}

```

cc @enrico-mi

david11133 committed:
Add initial Saildrones variables and rename mapping to params.py

david11133 committed:
Add Saildrones support with converter and update params

david11133 committed:
Improve fallback for missing temperature values using alternative column

david11133 committed:
implement converter script and add tests

enrico-mi commented:
Hi @david11133 ,

Good job. I have a few comments.

1. I just merged some [major updates](https://github.com/boom-lab/crocolaketools-public/pull/7) into `develop`. This affects your work but it shouldn't require many changes on your side. Basically the most important update for you is that I moved the configurations (e.g. the paths to the datasets) to a config.yaml file in the new config folder and therefore changed the way that `converter`s are initialized. Can you update the saildrone converter to follow the same convention? Let me know if it's not clear how it works with the update.
2. Similarly, it'd be great if saildrones2parquet could take a --config flag to read the arguments from the config file instead of the CLI. You can look at examples like glodap2parquet or spray2parquet that I just updated.
3. What do you think about changing params.py in crocolakeloader instead of crocolaketools/utils/params.py? I remember there was a good reason why you could not do it but if you could remind me why that'd be great, thanks!

Let me know if you have any questions!

david11133 commented:
Thanks @enrico-mi  for the heads-up!

I'll definitely update the Saildrone converter to follow the new convention, and I’ll adapt saildrones2parquet accordingly, no worries there.

Regarding params.py, I wasn’t able to move it to crocolakeloader because I don’t have write access to that repo. I’ll open a separate PR there to apply the changes and keep things clean.

Also, the recent updates make everything feel a lot cleaner and more modular, nice work!

david11133 committed:
i
i
Merge branch 'develop' into saildrones-converter

david11133 committed:
Add Saildrones configuration and update converter to use config.yaml

david11133 committed:
change the params in crocolakeloader

enrico-mi commented:
Great @david11133 and thanks for your feedback!

I believe `crocolakeloader` now is public, and you should be able to fork it and contribute back to it if that works for you?

I also fixed a bug in `converterSprayGliders` and pushed the changes. While this should not introduce conflicts in your PR, I also removed the old utils folder, and that is creating a conflict. It should be resolved once you update your tree with the `params` in `crocolakeloader`.

Let me know if you need any help with that.

david11133 commented:
Thanks @enrico-mi for the update!

However, I've updated the Saildrones converter to follow the new config-based structure and also modified saildrones2parquet to accept the --config flag as you suggested.

I'll resolve the conflict related to utils/params.py soon, sorry for the delay, I've just been a bit busy today.

Just a heads-up, I created a separate [PR (#1) ](https://github.com/boom-lab/crocolakeloader/pull/1) in crocolakeloader to apply the changes I made in params.py there. If there’s a better approach or something I’m missing, I’m happy to adjust my workflow going forward.

Thanks again, Enrico! Let me know if you spot any issues.

david11133 committed:
Remove old utils/params.py to resolve the conflict

enrico-mi commented on crocolaketools/converter/ConverterSaildrones.py:
```diff
@@ -0,0 +1,110 @@
+
+        qc_vars = ["TEMP", "PSAL", "PRES"]
+        if self.db_type == "BGC":
+            qc_vars = ["DOXY", "CHLA", "CDOM", "BBP700"]
```
We want to keep TEMP, PSAL, and PRES in the BGC dataset too, so you probably want to append to `qc_vars`  if the db_type is BGC?

enrico-mi commented on crocolaketools/test/test_converter.py:
```diff
@@ -1084,3 +1087,115 @@ def test_converter_cpr_convert(self):
+        assert "DOXY" in df.columns
+        assert "CHLA" in df.columns
+        assert "BBP700" in df.columns
+        assert "CDOM" in df.columns
```
You can add a check for the <PARAM>_QC values too

enrico-mi commented on scripts/saildrones2parquet.py:
```diff
@@ -0,0 +1,92 @@
+    converter = ConverterSaildrones(config=config)
+    converter.convert(saildrones_names)
+    del converter
+
```
fyi `del converter` is not need here, as we exit the function (and the whole script) right after. I usually use it when I have both conversions (PHY and BGC) in the same method, to make sure python knows that it can free up the memory before moving to the next

enrico-mi reviewed (changes requested):
Great @david11133 . I added a few minor comments.
I will test your PR on my machine too before merging it, probably on Wed or Thurs.

enrico-mi commented:
I have also pushed a new type of tests since you first opened your PR.

They test data integrity (see here: [crocolaketools/test/test_data.py](https://github.com/boom-lab/crocolaketools-public/blob/develop/crocolaketools/test/test_data.py)) by taking a random variable in some random netCDF file and looking for it in the converted parquet dataset. They're set up for Argo and SprayGliders at the  moment, but I tried to make the code general enough so that new datasets _should_ just require a few lines of codes. Of course there might always be specifications to each dataset that require tailored solutions.

Let me know if you want to give it a try and add the tests for the saildrones!

david11133 committed:
solve the minor issues in Saildrones converter

david11133 commented:
Thanks @enrico-mi for the feedback!

I’ve addressed the minor comments you mentioned in the code. The changes have been pushed, and everything is up to date now.

Additionally, I’ll be working on adding the Saildrones dataset to the `test_data` to test its integrity. I’ll keep you updated once that’s done.

Let me know if there's anything else you'd like me to adjust!

david11133 committed:
Resolved merge conflict in config.yaml

david11133 committed:
Resolved merge conflict in config.yaml

david11133 committed:
Merge remote-tracking branch 'upstream/develop' into saildrones-converter

david11133 commented:
Thanks @enrico-mi so much for the detailed feedback and for taking the time to walk me through all this, really appreciate your patience. I'll start working on the changes right away. I was actually in the middle of adding Saildrones to the test_data. Will keep you posted as I go!

Thanks again!

david11133 committed:
Replace hardcoded depths with dynamic values in read_to_df.

david11133 committed:
Update CLI arguments for consistency

david11133 committed:
Add entry point in setup.py and fix include_package_data  bug

david11133 committed:
Make --config flag boolean in saildrones2parquet.py

david11133 committed:
Rename ConverterSaildrones.py to converterSaildrones.py

enrico-mi commented on crocolaketools/converter/converterSaildrones.py:
```diff
@@ -0,0 +1,108 @@
+        if "latitude" not in df.columns or df["latitude"].isna().all():
+            raise ValueError("Latitude is missing or NaN in the dataset. Cannot compute pressure.")
+
+        df["PRES"] = gsw.p_from_z(-df["depth"], df["latitude"])
```
I would add a comment before this line to point out that depth values are expected negative by gsw as you need to swap the design with respect to the map that you defined. It's a small thing but it might make life easier for the next person implementing something similar or wondering how this is dealt with here (and a minus sign can always escape an attentive eye too!)

enrico-mi reviewed (changes requested):
Hi @david11133,

Well done. I looked through the changes and only found a typo that I commented, and I suggested adding a comment line in a place.

Next up for me is to test your new commits on my machine, which I will do tomorrow or Tuesday. What Saildrone files did you use for your tests? So that I use the same on my side.

About your questions:

> For O2_CONC_SBE37_MEAN (or DOXY) the NetCDF files list the unit as µmol/L, but the documentation mentions µmol/kg. Do you think we should convert to µmol/kg? If so, we’d need to incorporate seawater density to do that properly.

Good question. It happens that 1 kg (of water) and 1 l (of water) occupy the same volume, so µmol/kg and µmol/l are equivalent. It's true that this might not be 100% exact for seawater but I expect it to be equivalent as a unit of measure for oxygen density.

> For BKSCT_RED_MEAN, the long_name is "Optical backscatter at 650 nm" and the units are (m sr)^-1, which suggests this is BBP650 rather than BBP700. do we support BBP650 in this project?

Good observation. If BBP650 is not in the Argo parameter list, I would skip it for now.

david11133 committed:
switch Saildrone standardization test to use real NetCDF input

david11133 commented:
Thanks @enrico-mi for the helpful feedback!

I've added a comment to clarify the expected values for depth, and also fixed the typo in the test file, apologies for that, I had already corrected it locally but forgot to push the change. That should be updated now.

If you run into any issues testing the code on your machine, or if there’s anything you'd prefer to see handled differently, just let me know, happy to adjust as needed!

enrico-mi commented:
Great. Did you test it with some files from [here](https://www.pmel.noaa.gov/ocs/saildrone/data-access)? Which ones? Just so that I use the same

david11133 commented:
Oops, I just forgot to attach the files I used. Here they are:
`TPOS-2022_SD1052_1min.nc`,
`TPOS-2024_SD1033_1min.nc`,
`TPOS-2024_SD1090_1min.nc`,
`TPOS-2024_SD1030_1min.nc`.

david11133 committed:
tidy up code structure and naming

enrico-mi commented:
Hi @david11133 ,

Good job, the code works and executes successfully on my machine too now!

As it is, the script `saildrones2parquet` reads and converts one netCDF file from the input folder, throwing an error if multiple nc files are present in the input folder (it says that `read_to_ddf` is not implemented). The distinction between `read_to_df` and `read_to_ddf` is that the first one deals with pandas dataframe and works with one file only, while the second is used when the source contains multiple files (see `convert` method in `Converter` class). This is fine and it was meant to happen!

Ideally, when we convert a dataset (Saildrones, Argo, GOSHIP, etc) we want the script to read all the multiple files that are in the original dataset and convert them into one final parquet dataset. This is a next step that is more advanced. There are two ways to tackle it:

1. Serially: we loop over all the files in the folder and we apply all the transformations to each file, appending each resulting dataframe to a the previous ones
2. In parallel: the same but instead of looping we tell multiple threads to operate each on a file, and then combine the dataframes together.

I'd like to implement nb 2 as it is more efficient, more elegant, and in line with other converters. A good reference is `converterSpraygliders`. There you'll see that it has a few peculiarities:

- it uses `dask` to handle multithreading and larger-than-memory data (i.e. when combining dataframes together, it does so lazily and without loading all the data at once into memory, or the system would crash)
- it uses a "lock": this is needed because the underlying C library handling I/O operations for netCDF files is not thread-safe, in other words, more processes are allowed to access the same file. This is dangerous and redundant, so we use a lock as a semaphore, and each thread can read a file only when it "acquires" the lock (i.e. it has the green light, none else is accessing files)
- this requires a few extra "overhead" steps in handling the different passages (usually there is a layer where computations are prescribed, and a new upper layer where they're triggered, i.e. actually executed, and/or where are combined together).
- it also extra steps that Saildrones might not require, namely: re-chunking the data from each netCDF file, storing it temporarily to disk, and re-reading it before standardizing it. This is due to the fact that for large files, `dask` does not go smoothly from large xarray datasets to dask dataframes.

As I mentioned this is quite advanced: if you're interested in looking into this and learning about this, I am happy to mentor you. If not, I can address this myself and you can move ahead to writing documentation for the Saildrones.

What do you think?

david11133 commented:
Hey @enrico-mi ,

Thanks so much for the detailed feedback!

I’m definitely interested in moving to the next step and start supporting multiple file processing. The parallel approach sounds solid, as it lines up with the other converters. Plus, it’s a great learning opportunity for me.

While working on the Saildrones script, I actually spent some time digging into parallel solutions and also checked out your updates on the Spray Gliders converter, I think I’ve got a good grasp of how things are working under the hood now.

Quick question: do you think the Saildrones data is light enough that we can skip the whole chunking-and-reloading step? Like, are the files small/manageable enough to handle in memory without blowing things up? If you think some of them could get bulky, I’m totally fine with adding chunking, I was thinking we could chunk along the "row" dimension if we go that way.

I’ll start diving into this tomorrow after uni and keep you updated as I make progress. Let me know if there’s anything else I should keep in mind before jumping in!

Thanks again.

Best,
David

enrico-mi commented:
> Quick question: do you think the Saildrones data is light enough that we can skip the whole chunking-and-reloading step? Like, are the files small/manageable enough to handle in memory without blowing things up? If you think some of them could get bulky, I’m totally fine with adding chunking, I was thinking we could chunk along the "row" dimension if we go that way.

Looking at the saildrones data I see that the files are similar in size to the Spray Gliders', but that each variable is ~2 MB and we use very few of them (vs ~24 MB in some of the Spray Gliders). So I see two possible scenarios:

1. thanks to `invars` in `read_to_df` we are already subsetting a small enough dataset, and conversion to dataframe is fine and we don't need chunking and reloading
2. even if we are subsetting a small amount of data, that process is still demanding or not optimized for dask and it ends up loading too much data and we need to rechunking and reload. And you're correct in chunking along "row"!

I am hopeful that 1 works, and I'd start with that because I'd prefer not to write (and delete) extra files to disk, I think it can be a nuisance to the user and it's a smaller step for you. If it doesn't work, still those methods will be similar to the ones we'll use in solution 2, so it won't be wasted work.

david11133 committed:
Resolve conflicts with config

david11133 committed:
Resolve conflicts with test_converter

david11133 committed:
Revert accidental changes to unrelated files

david11133 committed:
Enable Saildrones converter to handle multiple files

david11133 committed:
Add Saildrones conf_cluster.yaml and improve Dask client handling in conversion script

david11133 committed:
Modified ConverterSaildrones to handle single and multiple file inputs more effectively

david11133 committed:
Update Saildrones converter to Prepare files before conversion

david11133 committed:
remove unused data preparation methods

david11133 committed:
Remove unnecessary repartitioning step

david11133 committed:
Enhance ConverterSaildrones for single file conversion

david11133 committed:
Refactor ConverterSaildrones to optimize single file conversion and improve Dask usage

david11133 committed:
Refactor ConverterSaildrones to ensure consistent column ordering and streamline single file processing

david11133 committed:
Remove unnecessary reindex from process_df()

david11133 committed:
Improve handling of single and multiple file processing

david11133 commented:
Thanks @enrico-mi  for the feedback!

About the overridden `convert()` method. I did that because the base class's `convert()` wasn't work well in the **single file** case, in this part:
```python
else:
   df = self.read_to_df(filenames[0],lock)
   if isinstance(df,pd.DataFrame):
      ddf = dd.from_pandas(df)
   elif isinstance(df,dd.DataFrame):
      ddf = df
```
…it expects `read_to_df()` to return a single DataFrame, but ConverterSaildrones’s `read_to_df` returns a delayed tuple (`df, invars`) (due to `@dask.delayed(nout=2)`). So, convert method fails the `isinstance` checks (`isinstance(df, pd.DataFrame)` or `isinstance(df, dd.DataFrame)`), leaving `ddf` uninitialized. But, it was working fine for multiple files since the base `convert()` uses `read_to_ddf()`, which handles things differently.

However, I found a cleaner workaround instead of overriding `convert()`. I used a class level flag (`self.is_multi_file`) to indicate whether multiple files are being processed. Since `read_to_ddf()` is only called for multiple files, set `self.is_multi_file = True` there. So the `read_to_df()` Check `self.is_multi_file` to decide whether to return only `df` or delayed tuple (`df, invars`). I think its much cleaner solution overall.

As for the `saildrones2parquet` script. thanks for pointing that out. I initially added the manual `config.yaml` loading to optimize single file runs (since Dask adds some overhead there), but the performance gain isn’t significant. I’ve removed that block and now rely on the converter’s internal handling, just like in `spray2parquet`.

Let me know what you think when you get a chance to test it again. Appreciate your time and help as always!

david11133 committed:
Add a per-file locks

david11133 committed:
Improve concurrency management

enrico-mi commented:
Thanks for the explanation @david11133 , I see why you overrode the function and it made sense. I'll test the latest version on Monday and I'll let you know if there is any change left to do.

david11133 committed:
Improve locking mechanism in ConverterSaildrones

david11133 committed:
Remove is-multi_file and add a parameter return_invars

david11133 committed:
Refactored convert() method for optimized single file processing

david11133 committed:
Modify the conversion calls in the script to use file paths

david11133 committed:
Combine the operations before computing in convert()

enrico-mi commented:
Hi @david11133 ,

I have tested your PR on my machine and also with a workflow that merges all the datasets together: it merges fine and the schemas are compatible!

I did find one thing that needs to be adjusted: there are some measurements that do not have valid pressure values. Basically, each ideally should have valid JULD, LATITUDE, LONGITUDE, and PRES: this allow the data user to always fully locate in time and space the measurement. Some data come with some pressure measurement to locate them, some with depth and we convert them, some we need to look at the metadata to learn where the sensor was installed.

When I filter by PRES, I find rows where PRES is \<NA\> but other measurements are valid, like TEMP or CHLA (see screenshot):

- I remember we discussed about temperature sensors' depth. Your code correctly has a map to assign the depth value, but the package that then converts it to pressure (`gsw`) also needs latitude, and latitude is \<NA\> sometimes, so we get rows with valid TEMP values but no valid PRES (nor LATITUDE). If the original data actually does not have valid latitude/longitude values, I suggest just discarding those measurements as we cannot locate them anyway.
- While looking into this, it seems to me that your code only maps TEMP_SBE37_MEAN to TEMP, not also TEMP_DEPTH_HALFMETER_MEAN (the latter seems to appear only in `depth_map`). If I am correct, then you should add this other mapping too. Note that you have two variables mapping into one dataframe column. This might turn out a bit of a pain to make it work, I'll think about it if I can come up with any idea to help you.
- For other measurements, like CHLA, it's not a matter of `gsw` rather that there is no mapping in `depth_map`. I gave a quick look to one of the original files and I could not find info about the chlorophyll sensor's depth. Can you please look into another few, and check also other variables, and add the missing mapping? It's plausible that they just don't have a depth assigned to them anywhere: please just make a list of all the variables/sensors for these, and I'll reach out to the PIs who gathered the data to see if they can help.

As always, let me know if you questions, comments, etc!

![image](https://github.com/user-attachments/assets/e58d091d-d557-40a2-bc1e-a40b19df73d3)

david11133 committed:
Enhance depth assignment to be based on metadata from installed sensors

david11133 committed:
Filter out rows where latitude or longitude are missing

david11133 committed:
Assign fixed depth values based on sensor metadata in ConverterSaildrones

david11133 committed:
Combine multi temperature sources into the TEMP column

david11133 committed:
Remove TEMP mapping from params and handle it in the code

david11133 committed:
Handle some files where TEMP_DEPTH_HALFMETER_MEAN does not exist

david11133 commented:
Hi @enrico-mi,

Just wanted to share a quick update on my work.

I’ve updated the converter to handle missing lat/lon values, now any row missing LATITUDE, LONGITUDE gets discarded, so all rows in the final dataset should have valid values for LATITUDE, LONGITUDE, JULD, or PRES.

I also added `TEMP_DEPTH_HALFMETER_MEAN` to the saildrones variable list. You can check it out in [PR #3](https://github.com/boom-lab/crocolakeloader/pull/3) on the `corocolakeloader` repo. The code now combines both `TEMP_SBE37_MEAN` and `TEMP_DEPTH_HALFMETER_MEAN` into `TEMP`.

I filled in the missing mappings in the `depth_map`, but I noticed something odd: some rows have valid values from sensors placed at different depths. like having values for both `TEMP_DEPTH_HALFMETER_MEAN` (0.5m) and `O2_CONC_SBE37_MEAN` (1.7m). That creates a bit of a conflict. cause which depth should we assign to the row? I think that's because multiple sensors may be installed at different depths, and sometimes their readings overlap in time.

Right now, I’m prioritizing the temperature sensor’s depth in those cases. Let me know if I am wrong or missing a thing. If thats a problem we could discuss and figure out a solution.

Whenever you get a chance, take a look at the latest updates and let me know your thoughts. Enjoy your weekend!

enrico-mi commented:
> I filled in the missing mappings in the `depth_map`, but I noticed something odd: some rows have valid values from sensors placed at different depths. like having values for both `TEMP_DEPTH_HALFMETER_MEAN` (0.5m) and `O2_CONC_SBE37_MEAN` (1.7m). That creates a bit of a conflict. cause which depth should we assign to the row? I think that's because multiple sensors may be installed at different depths, and sometimes their readings overlap in time.

Yes, that is expected. Ideally every row is a unique set of measurements, and the priority is to preserve as many measurements as we can. So in this case, assuming that for the same latitude, longitude and time we have valid measurements at different depths, what we want to do is to have two rows in the dataframe: a row with depth 0.5m, a valid temperature value, and \<NA\> for absent measurements (e.g. O2); and a row with depth 1.7m, a valid O2 and \<NA\> for variables for which we don't have a measurement. I guess one way of seeing it is that latitude, longitude, time and pressure (or depth) are the 4 coordinates of our system and if multiple measurements share the same set of [latitude, longitude, time and pressure] then we put them in the same row, if they don't we put them in different rows. Each measurements "requires" a set of [latitude, longitude, time and pressure], but not each set of [latitude, longitude, time and pressure] requires all measurements to be valid at that point, there can be \<NA\>s.

david11133 committed:
Implement depth-based row separation and reorder columns in single file

david11133 committed:
Tidy up code: update comments, and clean minor sections for clarity

david11133 commented:
Hi @enrico-mi,

I've been thinking about the profiling technique for the Saildrones data. I remember you mentioned that Saildrones don’t have classical profiling features, and we could just assign a specific cycle number to each mission.

From what I understand, that would mean assigning a fake but meaningful `PLATFORM_NUMBER` and `CYCLE_NUMBER` to each row based on the Saildrone’s mission. maybe using the filename or available metadata for the `PLATFORM_NUMBER`.

I'm still not entirely sure what would be a good approach for the `CYCLE_NUMBER`, though. whether we just keep it fixed (like always 1) or if there’s a better way to differentiate rows.

Let me know if that makes sense to you, and if you already have something in mind for how you'd like it to work, I'd love to hear more when you have time.

enrico-mi commented:
Hi @david11133 

> Would be awesome if you could try it out when you get a chance. Let me know if anything looks off or if you have suggestions!

I just tested it and it works, with no more issues of having null values for the dimensions (pres, lat, lon, juld)!

Fyi while looking at it I noticed a bug in `converter.py` in the method I coded to add flags: it assigns the same QC flag for all measurements, also when they are missing, which is not good. I'll open an issue.

> From what I understand, that would mean assigning a fake but meaningful PLATFORM_NUMBER and CYCLE_NUMBER to each row based on the Saildrone’s mission. maybe using the filename or available metadata for the PLATFORM_NUMBER.
> 
> I'm still not entirely sure what would be a good approach for the CYCLE_NUMBER, though. whether we just keep it fixed (like always 1) or if there’s a better way to differentiate rows.

Yes, you are correct from what I understand. Basically for all the rows coming from the same mission of the same saildrone (i.e. same file), we would assign the same PLATFORM_NUMBER and CYCLE_NUMBER. For PLATFORM_NUMBER we can use the WMO ID (it's in the datasets' `ds.attrs['wmo_id']` attribute). The WMO ID is a unique identifier that is assigned to ocean vehicles transmitting measurements over some network (see [here](https://community.wmo.int/en/buoy-wmo-identification-numbers)). For CYCLE_NUMBER, I don't see cycles or profiles per se in the data of each mission (there is only `row` dimension), but each drone has participated in multiple missions. There is a `mission` attribute in the datasets, but it's a wordy string, not a number (and we want CYCLE_NUMBER to be a number). In the `summary` of both the SD1033 and SD1090 from 2024, for example, they state it's mission 7, which coincides with the first mission being in 2017 having skipped 2020 (see [here](https://www.pmel.noaa.gov/ocs/saildrone/data-access)). Long story short: you can either kind-of-hardcode them or you can get the starting year of the mission, subtract 2017, and add 1 if it's the starting year is 2021 or larger.

Let me know what you think!

david11133 committed:
Add wmo_id and CYCLE_NUMBER calculations

david11133 commented:
Hi @enrico-mi ,

Just added the profiling feature for Saildrones today.

I used the `wmo_id` global attribute and assigned it to the `df`, then mapped it in params to `PLATFORM_NUMBER`. Also calculated `CYCLE_NUMBER` based on the mission start year, skipping 2020 as discussed. there’s a small change in crocolakeloader as well to correctly map the wmo_id. don’t forget to check it out [PR #4](https://github.com/boom-lab/crocolakeloader/pull/4).

This is my first time implementing a profiling feature in a converter, so I’m open in case you had a different approach or structure in mind. happy to adjust anything.

Also, let me know if there are any major updates or features still left to work on in the Saildrones converter!

On a side note: if you'd like me to fix the QC flag bug in `converter.py`, I can open a separate PR for that as well. But if you'd prefer to leave it for someone new (for onboarding or visibility), that works too. no worries either way.

Thanks a lot!

enrico-mi commented:
Hi @david,

The way you implemented the profile number looks fine to me.

On my end I tested the converter on all the saildrones files at once. There is an issue with the files from 2017 and 2018 as the converter fails for them. From a quick look at the .nc files, it looks like at least some variable names are different (for temperature for example). I'd recommend you to check them and in case match them, adapt the depth (let me know if it's not noted anywhere), etc. Sorry about this, it's the (ugly) reality of real data, where there are mismatches also in relatively small datasets...

Another note is the `read_to_df` method. Inside the try-except statement, you perform some computations on the dataframe `df` that do not depend on `ds` anymore. While they are not the most time-consuming operations in the whole converter, it is good practice to move them out of the try-except statement as you are using a lock and the other processes are waiting for the lock to be released to go read another file. An even more elegant solution would be to make the lock file-based, i.e. so that a process only needs to acquire a lock to read a file that is already being read by another process, not to read any netcdf file. This is quite challenging and when I looked into it I did not find a quick solution, but if you're up to exploring it, feel free to do it.

> On a side note: if you'd like me to fix the QC flag bug in converter.py, I can open a separate PR for that as well. But if you'd prefer to leave it for someone new (for onboarding or visibility), that works too. no worries either way.

how is your work load going so far? if you think you have time, it'd be great if you can tackle that!

david11133 committed:
Updated depth mapping for additional temperature variables and improve concurrency handling

david11133 commented:
> That would be great. Maybe we could have a table of this kind:

The table idea is great, but after checking all the files, I think the situation is a bit more generalized and probably doesn't need one. Here's what I found:

**Files from 2017–2019:**
I looked through all of them, and there’s no info about sensor depths at all. The only exception is some 2018 files that mention SBE56, but that variable doesn’t appear in the actual data variables. So, I don’t think it’s reported.

**Files from 2021–2024:**
These all explicitly include depths for SBE37 sensors, which is great. However, they’re missing depth info for chlorophyll, CDOM, and BKSCT sensors.

Also, just a quick note: while `installed_height` is technically provided per variable, as we discussed earlier, it's not always accurate, so I’m not relying on it unless confirmed.

> Make sure that the differences are by year, and not by file

Yep, I’ll make sure of that. But first we’ll need to gather the missing depth values. for the BGC sensors and files from 2017-2019, before finalizing anything year-specific.

By the way, let me know if we’re interested in variables from `UNCOR` or `RBR` sensors. I’m seeing stuff like `O2_CONC_UNCOR_MEAN` and `O2_RBR_CONC_MEAN` in some files. I remember you mentioned we care about `CTD`, but just wanted to double-check since some files include multiple versions of the same variable from RBR, CTD, and SBE37 (Like, `O2_CONC_SBE37_MEAN`, `O2_CONC_RBR_MEAN`).

david11133 commented:
Just extracted the depths info from all the files. in case you’re planning to reach out to the PIs.

I created two tables:

- One with depths pulled from each file’s `summary` attribute

- Another from the `installed_height` attribute of each variable

If you take a look, you’ll notice some differences between the two sources. The `installed_height` is present in all files (except for TPOS-2017_SD1005_1min.nc), but unfortunately, it’s not always reliable.


### **depths_from_summary**

| filename                 | TEMP_SBE37_MEAN | TEMP_CTD_MEAN | TEMP_DEPTH_HALFMETER_MEAN | SAL_SBE37_MEAN | SAL_MEAN | SAL_RBR_MEAN | O2_CONC_SBE37_MEAN | O2_CONC_MEAN | O2_CONC_UNCOR_MEAN | O2_RBR_CONC_MEAN | O2_CONC_RBR_MEAN | CHLOR_WETLABS_MEAN | CHLOR_MEAN | CHLOR_RBR_MEAN | CDOM_MEAN | BKSCT_RED_MEAN |
| ------------------------ | --------------- | ------------- | ------------------------- | -------------- | -------- | ------------ | ------------------ | ------------ | ------------------ | ---------------- | ---------------- | ------------------ | ---------- | -------------- | --------- | -------------- |
| TPOS-2017_SD1005_1min.nc |                 | ?             |                           |                | ?        |              |                    | ?            |                    |                  |                  |                    | ?          |                | ?         | ?              |
| TPOS-2017_SD1006_1min.nc |                 | ?             |                           |                | ?        |              |                    | ?            |                    |                  |                  |                    | ?          |                | ?         | ?              |
| TPOS-2018_SD1005_1min.nc |                 | ?             |                           |                | ?        |              |                    |              | ?                  |                  |                  |                    | ?          |                | ?         | ?              |
| TPOS-2018_SD1006_1min.nc |                 | ?             |                           |                | ?        |              |                    |              | ?                  |                  |                  |                    | ?          |                | ?         | ?              |
| TPOS-2018_SD1029_1min.nc |                 | ?             |                           |                | ?        |              |                    |              |                    | ?                |                  |                    |            | ?              |           |                |
| TPOS-2018_SD1030_1min.nc |                 | ?             |                           |                | ?        |              |                    |              |                    | ?                |                  |                    | ?          |                |           |                |
| TPOS-2019_SD1066_1min.nc | ?               |               |                           | ?              |          | ?            | ?                  |              |                    |                  | ?                | ?                  |            | ?              |           |                |
| TPOS-2019_SD1067_1min.nc | ?               |               |                           | ?              |          | ?            | ?                  |              |                    |                  | ?                | ?                  |            | ?              |           |                |
| TPOS-2019_SD1068_1min.nc | ?               |               |                           | ?              |          | ?            | ?                  |              |                    |                  | ?                | ?                  |            | ?              |           |                |
| TPOS-2019_SD1069_1min.nc | ?               |               |                           | ?              |          | ?            | ?                  |              |                    |                  | ?                | ?                  |            | ?              |           |                |
| TPOS-2021_SD1065_1min.nc | 1.54m           |               | 0.5m                      | 1.54m          |          |              | 1.54m              |              |                    |                  |                  | ?                  |            |                |           |                |
| TPOS-2021_SD1066_1min.nc | 1.54m           |               |                           | 1.54m          |          |              | 1.54m              |              |                    |                  |                  | ?                  |            |                |           |                |
| TPOS-2022_SD1033_1min.nc | 1.7m            |               | 0.5m                      | 1.7m           |          |              | 1.7m               |              |                    |                  |                  | ?                  |            |                |           |                |
| TPOS-2022_SD1052_1min.nc | 1.7m            |               | 0.5m                      | 1.7m           |          |              | 1.7m               |              |                    |                  |                  | ?                  |            |                |           |                |
| TPOS-2023_SD1030_1min.nc | 1.7m            |               | 0.5m                      | 1.7m           |          |              | 1.7m               |              |                    |                  |                  | ?                  |            |                |           |                |
| TPOS-2023_SD1033_1min.nc | 1.7m            |               | 0.5m                      | 1.7m           |          |              | 1.7m               |              |                    |                  |                  | ?                  |            |                |           |                |
| TPOS-2023_SD1079_1min.nc | 1.7m            |               | 0.5m                      | 1.7m           |          |              | 1.7m               |              |                    |                  |                  | ?                  |            |                |           |                |
| TPOS-2024_SD1033_1min.nc | 1.7m            |               | 0.5m                      | 1.7m           |          |              | 1.7m               |              |                    |                  |                  | ?                  |            |                | ?         | ?              |
| TPOS-2024_SD1090_1min.nc | 1.7m            |               | 0.5m                      | 1.7m           |          |              | 1.7m               |              |                    |                  |                  | ?                  |            |                | ?         | ?              |


### **depths_from_installed_height**

| filename                 | TEMP_SBE37_MEAN | TEMP_CTD_MEAN | TEMP_DEPTH_HALFMETER_MEAN | SAL_SBE37_MEAN | SAL_MEAN | SAL_RBR_MEAN | O2_CONC_SBE37_MEAN | O2_CONC_MEAN | O2_CONC_UNCOR_MEAN | O2_RBR_CONC_MEAN | O2_CONC_RBR_MEAN | CHLOR_WETLABS_MEAN | CHLOR_MEAN | CHLOR_RBR_MEAN | CDOM_MEAN | BKSCT_RED_MEAN |
| ------------------------ | --------------- | ------------- | ------------------------- | -------------- | -------- | ------------ | ------------------ | ------------ | ------------------ | ---------------- | ---------------- | ------------------ | ---------- | -------------- | --------- | -------------- |
| TPOS-2017_SD1005_1min.nc |                 | ?             |                           |                | ?        |              |                    | ?            |                    |                  |                  |                    | ?          |                | ?         | ?              |
| TPOS-2017_SD1006_1min.nc |                 | -0.6 m        |                           |                | -0.6 m   |              |                    | -0.6 m       |                    |                  |                  |                    | -0.25 m    |                | -0.25 m   | -0.25 m        |
| TPOS-2018_SD1005_1min.nc |                 | -0.6 m        |                           |                | -0.6 m   |              |                    |              | -0.6 m             |                  |                  |                    | -0.25 m    |                | -0.25 m   | -0.25 m        |
| TPOS-2018_SD1006_1min.nc |                 | -0.6 m        |                           |                | -0.6 m   |              |                    |              | -0.6 m             |                  |                  |                    | -0.25 m    |                | -0.25 m   | -0.25 m        |
| TPOS-2018_SD1029_1min.nc |                 | -0.53 m       |                           |                | -0.53 m  |              |                    |              |                    | -0.53 m          |                  |                    |            | -0.53 m        |           |                |
| TPOS-2018_SD1030_1min.nc |                 | -0.53 m       |                           |                | -0.53 m  |              |                    |              |                    | -0.53 m          |                  |                    | -0.53 m    |                |           |                |
| TPOS-2019_SD1066_1min.nc | -0.5 m          |               |                           | -0.5 m         |          | -0.53 m      | -0.5 m             |              |                    |                  | -0.53 m          | -0.5 m             |            | -0.53 m        |           |                |
| TPOS-2019_SD1067_1min.nc | -0.5 m          |               |                           | -0.5 m         |          | -0.53 m      | -0.5 m             |              |                    |                  | -0.53 m          | -0.5 m             |            | -0.53 m        |           |                |
| TPOS-2019_SD1068_1min.nc | -0.5 m          |               |                           | -0.5 m         |          | -0.53 m      | -0.5 m             |              |                    |                  | -0.53 m          | -0.5 m             |            | -0.53 m        |           |                |
| TPOS-2019_SD1069_1min.nc | -0.5 m          |               |                           | -0.5 m         |          | -0.53 m      | -0.5 m             |              |                    |                  | -0.53 m          | -0.5 m             |            | -0.53 m        |           |                |
| TPOS-2021_SD1065_1min.nc | -0.5 m          |               | -0.5 m                    | -0.5 m         |          |              | -0.5 m             |              |                    |                  |                  | -0.5 m             |            |                |           |                |
| TPOS-2021_SD1066_1min.nc | -0.5 m          |               |                           | -0.5 m         |          |              | -0.5 m             |              |                    |                  |                  | -0.5 m             |            |                |           |                |
| TPOS-2022_SD1033_1min.nc | -0.5 m          |               | -0.5 m                    | -0.5 m         |          |              | -0.5 m             |              |                    |                  |                  | -0.5 m             |            |                |           |                |
| TPOS-2022_SD1052_1min.nc | -0.5 m          |               | -0.5 m                    | -0.5 m         |          |              | -0.5 m             |              |                    |                  |                  | -0.5 m             |            |                |           |                |
| TPOS-2023_SD1030_1min.nc | -1.7 m          |               | -0.5 m                    | -1.7 m         |          |              | -1.7 m             |              |                    |                  |                  | -1.9 m             |            |                |           |                |
| TPOS-2023_SD1033_1min.nc | -1.7 m          |               | -0.5 m                    | -1.7 m         |          |              | -1.7 m             |              |                    |                  |                  | -1.9 m             |            |                |           |                |
| TPOS-2023_SD1079_1min.nc | -1.7 m          |               | -0.5 m                    | -1.7 m         |          |              | -1.7 m             |              |                    |                  |                  | -1.9 m             |            |                |           |                |
| TPOS-2024_SD1033_1min.nc | -1.7 m          |               | -0.5 m                    | -1.7 m         |          |              | -1.7 m             |              |                    |                  |                  | -1.9 m             |            |                | -1.9 m    | -1.9 m         |
| TPOS-2024_SD1090_1min.nc | -1.7 m          |               | -0.5 m                    | -1.7 m         |          |              | -1.7 m             |              |                    |                  |                  | -1.9 m             |            |                | -1.9 m    | -1.9 m         |

david11133 committed:
Add depths for sensors

david11133 committed:
remove unimportant sensors

david11133 committed:
Merge branch 'develop' into saildrones-converter

enrico-mi commented:
Hi @david11133 ,

Thanks for the detailed response.

> By the way, let me know if we’re interested in variables from UNCOR, AANDERAA or RBR sensors. I’m seeing stuff like O2_CONC_UNCOR_MEAN, O2_RBR_CONC_MEAN, O2_CONC_AANDERAA_MEAN in some files. I remember you mentioned we care about CTD, but just wanted to double-check since some files include multiple versions of the same variable from RBR, AANDERAA, and SBE37 (Like, O2_CONC_SBE37_MEAN, O2_CONC_AANDERAA_MEAN, O2_CONC_RBR_MEAN).

Yes, generally speaking we want the data also from different sensors (similarly to what we do with temperature). I had used one of the recent files when I mentioned O2_CONC_SBE37_MEAN and it did not have other O2 concentration measurements, but if other files have them, please include them (taking care of different depths like for temperature, etc.).
The list of all the variables we want is [here](https://crocolakedocs.readthedocs.io/en/latest/crocolake.html#crocolake-s-conventions). Generally speaking, we'd like to preserve all data that matches that table, no matter what sensor or depth it comes from (as we did for temperature). Discard any measurement that does not match a variable already in the table (for example current velocity) and any measurement that is *not* of the ocean (e.g. *air* temperature). Feel free to ask any time you have a doubt!

> Just extracted the depths info from all the files. 

That's fantastic, thanks, I'll contact them.
For now, I suggest that you proceed like this:

1. if no depth info is available, discard the measurement
2. if depth is available only from one of summary and installation_height, use that value
3. if depth is indicated in both summary and installation_height:
    a. if it's consistent use it
    b. if it's inconsistent, use the one from summary (it seems more plausible to me)

This is what we can do for now. If we get new information we will open an issue and do a separate PR.

david11133 committed:
Add more sensors and handle multi sensors per file

david11133 committed:
resolve conflict in config file

david11133 committed:
Resolve conflict with the default relative path for demo data

david11133 committed:
Ensure all identifiers are non-null

david11133 committed:
restart the server between PHY and BGC

david11133 commented:
Hi @enrico-mi ,

> Yes, generally speaking we want the data also from different sensors (similarly to what we do with temperature). 

Perfect, I’ve updated the code accordingly and added all the sensors I found across the dataset (see [PR #5](https://github.com/boom-lab/crocolakeloader/pull/5)). I also updated the process_df() logic to handle cases where multiple sensors contribute to the same column (like `TEMP`)

> For now, I suggest that you proceed like this:

>1. if no depth info is available, discard the measurement
>2. if depth is available only from one of summary and installation_height, use that value
?3. if depth is indicated in both summary and installation_height:
>   a. if it's consistent use it
>   b. if it's inconsistent, use the one from summary (it seems more plausible to me)

That’s exactly how I’ve handled it for now. I’ll keep it like this until we get confirmation of the real depth values, then we can decide if we want to shift strategies.

Also, I came across a few variables in the data like:

- `SW_UNMASKED_IRRAD_CENTER_MEAN`
- `LW_IRRAD_MEAN`
- `SW_IRRAD_DIFFUSE_MEAN`
- `SW_IRRAD_TOTAL_MEAN`
- `SW_UNMASKED_IRRAD_6DET_MEAN`

Do you think these might be related to variables like `DOWN_IRRADIANCE380`, `DOWN_IRRADIANCE412`, etc.? Just wanted to check in case they should be considered part of the BGC group.

Anyway, once you get a chance to try the updated code, let me know if you have any feedback or updates.

david11133 committed:
Optimize Depth Annotation and Column Merging

david11133 committed:
combine columns using pandas' combine_first method

david11133 committed:
combine sensors with the same depth in one row

david11133 commented:
Also, just wanted to give you a heads-up about a performance issue with ConverterSaildrones that you might’ve already noticed.

When processing big files or a large number of files like the full Saildrones dataset. things tend to slow down a lot. You might see warnings about graph size and multiple shuffle operations. Thats mainly because Saildrones data requires some heavier processing, like `groupby` in read_to_df() and Column merging in `process_df()`

These steps cause the data to be redistributed across workers, which leads to high network and disk I/O. Plus, `groupby` and `concat` are memory-intensive and naturally trigger Dask shuffling. It gets especially intense when processing multiple files since each file produces a large intermediate DataFrame.

You can really feel it during BGC conversion, since it handles more variables (like `DOXY`, `CHLA`, `CDOM`, `BBP700`), which just adds to the DataFrame size and task graph complexity.

Since the goal is to collect as many quality measurements as we can from the Saildrones data. So the processing step is doing a lot of work upfront.

I’ll definitely keep looking into optimizations. But for now, let me know if there are any major updates or changes we should prioritize before the first evaluation deadline.

enrico-mi commented:
> I think you might be working with an older version of the code. I just tested the latest version and it’s working for me:

yes, I forgot to merge a crocolakeloader PR and now it all works, thanks!

> Also, just wanted to give you a heads-up about a performance issue with ConverterSaildrones that you might’ve already noticed.

Thank you for the heads up. I wanted to bring this up, but the first goal was to have a converter work successfully, and optimizing the graph size is not a trivial task.

I see that sometimes the graph is also >100MB large, which is definitely a lot. In my experience, vectorizing operations helps. For example, I had one case with `converterArgoQC` where I was applying `keep_best_values` with `map_partition`, but I was applying it while looping over each column. By looping _inside_ `map_partition` (i.e. calling map_partition once, and looping over the columns inside `keep_best_values`) instead of calling `map_partition` multiple times inside one loop, the graph size dropped significantly. I believe it's because the graph needs to take track of _one_ operation (inside `map_partition` we're operating on a pandas df, and the `meta` argument of `map_partition` lets dask know what to expect back to keep building the graph), instead of one per column like when I was calling `map_partition` inside the loop.

What do you think of trying to optimize the code and, if you can't find a way to reduce the graph size by Wednesday or so, we merge what you currently have?

david11133 committed:
Vectorize depth assignment and move it to process_df()

david11133 committed:
Chunk large dataframes to manage memory

enrico-mi commented:
Hi @david11133 

All works fine, also the graph size is much smaller.

I'm merging the PR, congrats, this was quite a big effort!

We will need to add tests (for data integrity at least), but I'll open an issue for that.
